### PR TITLE
*: Respect route replacement mode in TrafficPolicy plugin  

### PR DESCRIPTION
**Pull request by timflannagan**
# Description

This is the initial PR that implements the [route replacement](https://github.com/kgateway-dev/kgateway/pull/11227) EP. These changes focus on proactively catching structurally valid but semantically invalid configurations that Envoy would NACK at runtime.

From the 1.x APIs, the transformation policy type is an excellent example where this type of guardrail is critical. While a policy might be structurally valid (correct YAML, valid fields), it could contain semantically invalid configurations like:

- Malformed template expressions
- Template references to non-existent headers or variables
- Invalid body transformations that would produce malformed JSON/XML

The TP plugin has been updated to perform additional validation checks depending on the route replacement mode that's been configured for the KG installation:

- PGV validation: Runs in *standard* mode, performs quick validation of Envoy configuration types performs basic structural and semantic validation. This short circuits expensive xDS validation when basic issues are found, such as invalid header values or malformed template syntax.
- xDS validation (via Envoy validate mode): Runs in *strict* mode and controls whether the plugin will convert Envoy types stored in the IR into a partial bootstrap config and passes this to Envoy in validate mode. This detects potential NACKs before they happen at runtime and catches complex semantic issues like invalid transformations.

Whenever either of these validation checks fail, the route translator replaces the invalid route with a 500 direct response action. Previously, errors encountered during attached policy IR construction only influenced the policy status, which led to valid routes with invalid policy attached to them being pushed to the xDS snapshot and risk undefined behavior, fail open scenarios w.r.t security policy, or NACKs from the proxy. Now, the route translator has been updated to respect these policy IR errors, and trigger route replacement, in addition to the existing policy status.

Additionally, several new libraries have been introduced with these changes:

- Dedicated Envoy bootstrap library. Responsible for providing a generic partial bootstrap builder
- Abstracting envoy validate mode via an xDS validation library

These abstractions will be used in follow-up work that builds on top of this foundation and performs additional validation checks (e.g. CDS, RDS, etc.) that help prevent potential NACKs from reaching the fleet of managed proxies.

# Change Type

/kind feature

# Changelog

```release-note
The TrafficPolicy plugin now respects the route replacement mode setting (`KGW_ROUTE_REPLACEMENT_MODE`). When in strict mode, the plugin performs additional validation to catch invalid configurations before they reach Envoy. Invalid policies that would cause Envoy to NACK at runtime (e.g. malformed templates) will now be replaced with a direct response (HTTP 500) and report clear status conditions. This prevents fail-open scenarios where invalid policies could allow unintended traffic.
```

# Additional Notes

Fixes https://github.com/kgateway-dev/kgateway/issues/11535

### CODE REVIEW

File: internal/kgateway/translator/gateway/gateway_translator_test.go [hunk 1]
```diff
@@ -746,17 +746,175 @@ var _ = DescribeTable("Basic GatewayTranslator Tests",
 				assertPolicyStatusWithGeneration(reportsMap, expectedPolicies, 42)
 			},
 		}),
-	// TODO: Add this once istio adds support for listener sets
```
Comment (timflannagan) [inline internal/kgateway/translator/gateway/gateway_translator_test.go:862]: TODO: revert.

File: pkg/pluginsdk/ir/iface.go [hunk 2]
```diff
@@ -55,6 +60,23 @@ func (r *TypedFilterConfigMap) GetTypedConfig(key string) proto.Message {
 	return nil
 }
 
+func (r *TypedFilterConfigMap) ToAnyMap() map[string]*anypb.Any {
```
Comment (timflannagan) [inline pkg/pluginsdk/ir/iface.go:63]: TODO: Adopt this method throughout the route translator.

File: internal/kgateway/translator/irtranslator/route.go [hunk 3]
```diff
@@ -210,6 +210,11 @@ func (h *httpRouteConfigurationTranslator) envoyRoutes(
 			out.Action = &envoy_config_route_v3.Route_DirectResponse{
```
Comment (timflannagan) [inline internal/kgateway/translator/irtranslator/route.go:207]: TODO: Strip the typedFilterConfig map.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	"context"
+
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/kgateway-dev/kgateway/v2/api/v1alpha1"
+	"github.com/kgateway-dev/kgateway/v2/pkg/validator"
+	"github.com/kgateway-dev/kgateway/v2/pkg/xds/bootstrap"
+)
+
+func (p *TrafficPolicy) Validate(ctx context.Context, v validator.Validator, policy *v1alpha1.TrafficPolicy) error {
```
Comment (timflannagan) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: Note: I kept the status optimization outside of this PR. There's some weirdness that I still need to figure out locally, and was leading to policy status being wiped / incorrectly reported on controller restarts.

**Comment by timflannagan** [comment]
TODO: Valid test cases for each policy type? Need a way to ensure that the validate.go is producing the correct config (ran into this yesterday with ext proc, ext auth, rate limit, etc) or refactor the TP plugin to ensure that we're sharing the same function between the validation and policy application code.


TODO: Added several test cases that catch errors during the IR Translate method. We need to handle these different than the validation checks w.r.t. when they trigger route replacement.

TODO: Determine whether every policy type within the TP API should perform PGV/xDS validation.

TODO: Validate the ext auth context extensions edge case with null byte. Was working and Envoy NACK'd but no longer reproducible locally.

**Comment by timflannagan** [comment]
TODO: Determine whether every policy type within the TP API should perform PGV/xDS validation.

File: internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go [hunk 5]
```diff
@@ -216,6 +219,13 @@ func NewPlugin(ctx context.Context, commoncol *common.CommonCollections) extensi
 		}
 
 		policyIR, errors := translator.Translate(krtctx, policyCR)
+		switch commoncol.Settings.RouteReplacementMode {
+		case settings.RouteReplacementStrict:
+			if err := policyIR.Validate(ctx, v, policyCR); err != nil {
+				logger.Error("strict mode validation failed", "policy", policyCR.Name, "error", err)
+				errors = append(errors, err)
+			}
+		}
```
Comment (timflannagan) [inline internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go]: TODO: Delineate between IR translate & validation errors.

File: internal/kgateway/translator/irtranslator/route.go [hunk 6]
```diff
@@ -315,8 +304,7 @@ func (h *httpRouteConfigurationTranslator) runRoutePlugins(
 			TypedFilterConfig: typedPerFilterConfig,
 		}
 		for _, pol := range mergePolicies(pass, pols) {
-			// TODO: should we append pol.Error to errs?
-			// i.e. errs = append(errs, pol.Error)
```
Comment (timflannagan) [inline internal/kgateway/translator/irtranslator/route.go:322]: TODO: Determine whether we need to respect pol.Errors in ComputeRouteConfiguration, runVhostPlugins, runBackendPolicies/runBackend.

We still use the pol.Errors when setting initial status in the runVhostPlugins method, while the translateRouteAction handles the errors from the runBackend* methods, but only logs them.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/kgateway-dev/kgateway/v2/api/v1alpha1"
+	"github.com/kgateway-dev/kgateway/v2/internal/kgateway/wellknown"
+	"github.com/kgateway-dev/kgateway/v2/pkg/validator"
+	"github.com/kgateway-dev/kgateway/v2/pkg/xds/bootstrap"
+)
+
+func (p *TrafficPolicy) Validate(ctx context.Context, v validator.Validator, policy *v1alpha1.TrafficPolicy) error {
+	if err := p.validateProto(ctx); err != nil {
+		return err
+	}
```
Comment (lgadban) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: I thought we agreed that PGV validation would NOT be gated by strict mode?

File: pkg/validator/validator.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+)
+
+const (
+	envoyPath  = "/usr/local/bin/envoy"
+	envoyImage = "quay.io/solo-io/envoy-gloo:1.34.1-patch1"
```
Comment (lgadban) [inline pkg/validator/validator.go]: not a blocker, but assuming we have a plan to keep this in sync with the correct version? linker version etc?

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+func (p *TrafficPolicy) validateProto(ctx context.Context) error {
+	// TODO: rustformations, and ext auth/rate limit provider validation
+	// Note: no need for buffer validation as it's a single int field, right?
+	var validators []func() error
+	if p.spec.AI != nil {
+		if p.spec.AI.Transformation != nil {
+			validators = append(validators, p.spec.AI.Transformation.Validate)
+		}
```
Comment (lgadban) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:49]: I'm a bit confused by this pattern; is the idea that all sub-policy IRs will add a `Validate` function eventually?
And the near-term plan is that if the sub-policy IR is a direct proto this will be handled already?

Comment (lgadban) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:49]: also, do we not need to nil check the `Validate` function ?

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	if err := p.validateProto(ctx); err != nil {
+		return err
+	}
+	if err := p.validateXDS(ctx, v); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *TrafficPolicy) validateProto(ctx context.Context) error {
```
Comment (lgadban) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: So the responsibility of the plugin modifier is to add any new IRs here manually?
And reviewer should help enforce this?

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	return nil
+}
+
+// validateXDS builds a partial bootstrap config and validates it via
+// envoy validate mode.
+func (p *TrafficPolicy) validateXDS(ctx context.Context, v validator.Validator) error {
+	builder := bootstrap.New()
+	if p.spec.transform != nil {
+		builder.AddFilterConfig(transformationFilterNamePrefix, p.spec.transform)
```
Comment (lgadban) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: assuming it is prohibitively challenging to re-use the e.g. ApplyForRoute method here,
i.e. we will have to write this "one off" code for the validation use case explicitly?

File: pkg/xds/bootstrap/builder.go [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+
+// AddRouteConfig adds a route configuration to the builder.
+func (b *ConfigBuilder) AddRouteConfig(r *envoy_config_route_v3.Route) {
+	b.routes = append(b.routes, r)
+}
+
+// AddClusterConfig adds a cluster configuration to the builder.
+func (b *ConfigBuilder) AddClusterConfig(c *envoy_config_cluster_v3.Cluster) {
+	b.clusters = append(b.clusters, c)
+}
```
Comment (lgadban) [inline pkg/xds/bootstrap/builder.go]: these seem to be unused currently, assuming usage will come eventually?

**Review by lgadban — COMMENTED** [review]
first pass, mostly LGTM but I have a few comments & questions

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 7 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: Oh I misunderstood since I had the opposite impression from previous conversations. It's such a light weight check that I'm not opposed to moving to STANDARD mode.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 9 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:49]: Yeah exactly - that's one of the TODOs where I wanted to sync with you. This implementation is very awkward in it's current shape imo. It's a minor cleanup change to avoid calling Validate() directly and delaying until all the PGV validator methods have been registered first.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 10 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: Another example I wanted to sync with you on this approach. High risk of maintenance with this approach imo.

File: pkg/xds/bootstrap/builder.go (hunk 12 above)
Comment (timflannagan) [reply pkg/xds/bootstrap/builder.go]: Right - currently unused, relevant to CDS/BackendConfigPolicy validation. Let's rip it out until it's useful.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 9 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:49]: Talked with lawrence offline - we're looking at cleaning up PGV-based validation, but it requires a traffic policy refactor (see #11353 for one approach, you'd add a Validate method to IRs in that pattern, etc.). Going to investigate this in a follow-up.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 10 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: Talked with lawrence offline - a couple of different ways we can handle this:

- Define a fake translation pass and call ApplyForRoute
- Determine whether we can make handleX pure functions instead of methods and re-use those functions

Leaning towards the former if it works / there isn't any obvious gaps with the approach.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 11 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go]: added this as an option to the other review comment: https://github.com/kgateway-dev/kgateway/pull/11553#discussion_r2192970641. marking as resolved to avoid too many threads.

File: internal/kgateway/translator/irtranslator/route.go (hunk 6 above)
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/route.go:322]: Talked with lawrence offline - https://github.com/kgateway-dev/kgateway/issues/10625#issuecomment-3049505602 is the summary.

File: internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go (hunk 5 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go]: Still relevant - will be done in a follow-up.

File: internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go [hunk 13]
```diff
@@ -216,6 +219,18 @@ func NewPlugin(ctx context.Context, commoncol *common.CommonCollections) extensi
...
+			if err := policyIR.validateStandard(); err != nil {
+				logger.Error("standard mode validation failed", "policy", policyCR.Name, "error", err)
+				errors = append(errors, err)
+			}
+		case settings.RouteReplacementStrict:
+			if err := policyIR.validateStrict(ctx, v); err != nil {
+				logger.Error("strict mode validation failed", "policy", policyCR.Name, "error", err)
+				errors = append(errors, err)
+			}
+		}
```
Comment (timflannagan) [inline internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go]: Open to defining a single Validate method that takes a settings.Settings parameter to determine which validation function to run.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+	if err := p.validateStandard(); err != nil {
+		return err
+	}
+	return p.validateXDS(ctx, v)
+}
+
+// validateProto performs basic proto validation that runs in STANDARD mode.
+// TODO(tim): this logic will be refactored in the future to be less brittle,
+// easier to read/maintain/etc. but requires additional traffic policy plugin
+// refactoring to do properly.
```
Comment (timflannagan) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:41]: Added a TODO for this code. In it's current form, it's incredibly hard to read and has a high risk of maintenance burden without a proper refactor that will be done in a follow-up.

File: internal/kgateway/translator/gateway/testutils/inputs/route-replacement/standard/gateway-level-policy-validation.yaml
Comment (timflannagan) [inline internal/kgateway/translator/gateway/testutils/inputs/route-replacement/standard/gateway-level-policy-validation.yaml:1]: Note: this test case and the listener-level attachment are no-ops. Follow-up work needs to be done to the route translator to handle higher-level attachment w.r.t. triggering route replacement.

File: internal/kgateway/translator/gateway/testutils/inputs/route-replacement/standard/invalid-httproute-prefix-match.yaml
Comment (timflannagan) [inline internal/kgateway/translator/gateway/testutils/inputs/route-replacement/standard/invalid-httproute-prefix-match.yaml:1]: Existing test case moved to the route replacement suite. Related to route replacement where the validateEnvoyRoute function will return an error, and that error triggers the output Envoy route to use a direct response action.

File: internal/kgateway/translator/irtranslator/route.go [hunk 15]
```diff
@@ -318,15 +307,16 @@ func (h *httpRouteConfigurationTranslator) runRoutePlugins(
...
+			// skip plugin application if we encountered any errors while constructing
+			// the policy IR.
```
Comment (timflannagan) [inline internal/kgateway/translator/irtranslator/route.go:311]: Just to highlight this behavior - this mainly has impact on the listener-level config that's disabled by default:

```diff
diff --git a/internal/kgateway/translator/gateway/testutils/outputs/route-replacement/strict/invalid-csrf-regex-config-out.yaml b/internal/kgateway/translator/gateway/testutils/outputs/route-replacement/strict/invalid-csrf-regex-config-out.yaml
index 8526acb691..f042e46727 100644
--- a/internal/kgateway/translator/gateway/testutils/outputs/route-replacement/strict/invalid-csrf-regex-config-out.yaml
+++ b/internal/kgateway/translator/gateway/testutils/outputs/route-replacement/strict/invalid-csrf-regex-config-out.yaml
@@ -23,11 +23,6 @@ Listeners:
       typedConfig:
         '@type': type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
         httpFilters:
-        - name: envoy.filters.http.csrf
-          typedConfig:
-            '@type': type.googleapis.com/envoy.extensions.filters.http.csrf.v3.CsrfPolicy
-            filterEnabled:
-              defaultValue: {}
```

File: pkg/pluginsdk/ir/iface.go [hunk 16]
```diff
@@ -55,6 +60,23 @@ func (r *TypedFilterConfigMap) GetTypedConfig(key string) proto.Message {
...
+	typedPerFilterConfigAny := map[string]*anypb.Any{}
+	for k, v := range *r {
+		if anyMsg, ok := v.(*anypb.Any); ok {
+			typedPerFilterConfigAny[k] = anyMsg
+			continue
+		}
+		config, err := utils.MessageToAny(v)
+		if err != nil {
+			logger.Error("unexpected marshalling error", "error", err)
```
Comment (timflannagan) [inline pkg/pluginsdk/ir/iface.go:72]: Existing error handling that I moved to this package. Open question on whether it's correct.

File: pkg/pluginsdk/ir/iface.go [hunk 17]
```diff
@@ -133,7 +155,8 @@ type ProxyTranslationPass interface {
 	ApplyForRoute(
 		ctx context.Context,
 		pCtx *RouteContext,
-		out *envoy_config_route_v3.Route) error
+		out *envoy_config_route_v3.Route,
+	) error
```
Comment (timflannagan) [inline pkg/pluginsdk/ir/iface.go:159]: Unrelated to these changes, but updated this formatting to be consistent with the other ApplyForX signatures.

File: pkg/xds/bootstrap/builder.go
Comment (timflannagan) [inline pkg/xds/bootstrap/builder.go:1]: Follow-up work will expand this library to be able to perform RDS/CDS validation. Good enough for now though.

File: internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go (hunk 13 above)
Comment (shashankram) [reply internal/kgateway/extensions2/plugins/trafficpolicy/traffic_policy_plugin.go]: I would prefer this - collapsing the check into policyIR.validate(commoncol.Settings.RouteReplacementMode)

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+func (p *TrafficPolicy) validateProto() error {
+	// TODO: rustformations, and ext auth/rate limit provider validation
+	// Note: no need for buffer validation as it's a single int field, right?
+	var validators []func() error
+	if p.spec.AI != nil {
+		if p.spec.AI.Transformation != nil {
+			validators = append(validators, p.spec.AI.Transformation.Validate)
```
Comment (shashankram) [inline internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:48]: why do we need per sub-IR Validate methods? validateProto() is an odd name, can this just be merged into `validateStandard()`?

File: internal/kgateway/translator/irtranslator/route.go [hunk 19]
```diff
@@ -208,6 +207,11 @@ func (h *httpRouteConfigurationTranslator) envoyRoutes(
 			out.Action = &envoy_config_route_v3.Route_DirectResponse{
 				DirectResponse: &envoy_config_route_v3.DirectResponseAction{
 					Status: http.StatusInternalServerError,
+					Body: &envoy_config_core_v3.DataSource{
+						Specifier: &envoy_config_core_v3.DataSource_InlineString{
+							InlineString: `kgateway detected an invalid route configuration and replaced it with a direct response.`,
```
Comment (shashankram) [inline internal/kgateway/translator/irtranslator/route.go]: nit: kgateway is implicit
```suggestion
							InlineString: `invalid route configuration detected and replaced with a direct response`,
```

File: pkg/validator/validator.go (hunk 8 above)
Comment (shashankram) [inline pkg/validator/validator.go]: Can this be set via Settings. I don't like hardcoding images in code

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 18 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:48]: @shashankram This is awkward in it's current form, but I didn't want to commit to a proper refactor in this PR. Open to suggestions on the name of this methods as well - it's just performing PGV validation for any envoy filter types that live on the TP IR.

File: pkg/validator/validator.go (hunk 8 above)
Comment (timflannagan) [reply pkg/validator/validator.go]: No disagreement - let me play around with the global settings approach since this image already skewed from the ENVOY_IMAGE Makefile variable. Even with the settings approach, I think we'll have to define this image in multiple places since it's used in the component Dockerfile(s). Additionally, it would introduce precedence where other data plane related images (e.g. SDS) need to surface as settings variables.

Comment (timflannagan) [reply pkg/validator/validator.go]: Attempted to fix https://github.com/kgateway-dev/kgateway/pull/11553/commits/0241637803d2798ad2612093a3bc3af3172fcbd0 using the linker flag approach.

Comment (timflannagan) [reply pkg/validator/validator.go]: yep good call - let's talk further in the https://github.com/kgateway-dev/kgateway/pull/11553#discussion_r2193381808 comment thread. resolving

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 18 above)
Comment (shashankram) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:48]: I was trying to understand the pattern. Can it just be inlined vs needing to implement a validate per sub-policy?

Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:48]: I’m worried about the maintenance cost wrt review burden if we didn’t formalize this somehow. In its current shape, it’s hard to read but we can standardize on the right approach going forward to minimize that cost

File: hack/utils/settings/get_envoy_image.go
Comment (timflannagan) [inline hack/utils/settings/get_envoy_image.go:1]: I'm not entirely sold on this approach, but we need the ENVOY_IMAGE Makefile variable as the docker images need that to be supplied as a build variable (and goreleaser leverages that variable too). I wasn't able to find a decent way to avoid defining the envoy image N times without breaking devs that run suites directly, e.g. `go tool ginkgo run -v -fail-fast internal/kgateway/translator/gateway`, without the corresponding ldflags provided.

One option includes providing a fallback envoy image in the pkg/envoy library for the `go test` edge cases. That seemed awkward and could become another issue where local & CI accidentally test with different images over time.

Comment (timflannagan) [reply hack/utils/settings/get_envoy_image.go:1]: I can also back this change and handle in a follow-up/track in the parent epic.

File: internal/kgateway/extensions2/plugins/trafficpolicy/validate.go (hunk 18 above)
Comment (timflannagan) [reply internal/kgateway/extensions2/plugins/trafficpolicy/validate.go:48]: Just to close the loop - https://github.com/timflannagan/kgateway/commit/4b16aad90e8ec730bec8df5be94f84ff73c1a419 is what I had in mind for one approach to make this more maintainable over time.

File: hack/utils/settings/get_envoy_image.go
Comment (shashankram) [reply hack/utils/settings/get_envoy_image.go:1]: > go tool ginkgo run -v -fail-fast internal/kgateway/translator/gateway, without the corresponding ldflags

Yes this is annoying. Let's just hardcode for now and come up with a better approach later. Could you add a note in the Makefile where the Envoy version is defined to also update this file that hardcodes the version again

File: pkg/envoy/image.go [hunk 20]
```diff
@@ -0,0 +1,8 @@
+package envoy
+
+// DefaultImage is the default Envoy proxy image used by the controller.
+const DefaultImage = "quay.io/solo-io/envoy-gloo:1.34.1-patch3"
+
+// Image is the Envoy proxy image used by the controller.
+// This is set by the linker during build.
+var Image = DefaultImage
```
Comment (shashankram) [inline pkg/envoy/image.go]: If we are hardcoding this anyway, what's the point of having an additional linker flag? I'd just get rid of the linker flag with the image hardcoded here, even if it is meant to serve as a fallback.

File: pkg/validator/validator.go [hunk 21]
```diff
@@ -0,0 +1,107 @@
...
+
+type dockerValidator struct {
+	img string
+}
+
+var _ Validator = &dockerValidator{}
+
+func (d *dockerValidator) Validate(ctx context.Context, yaml string) error {
+	cmd := exec.CommandContext(ctx,
+		"docker", "run",
```
Comment (shashankram) [inline pkg/validator/validator.go:64]: Where is the dockerValidator used?

File: pkg/validator/validator.go [hunk 22]
```diff
@@ -0,0 +1,107 @@
+package validator
```
Comment (shashankram) [inline pkg/validator/validator.go:1]: could you add unit tests in validator_test.go

File: pkg/xds/bootstrap/builder.go [hunk 23]
```diff
@@ -0,0 +1,96 @@
+package bootstrap
```
Comment (shashankram) [inline pkg/xds/bootstrap/builder.go:1]: needs unit tests

File: internal/kgateway/translator/irtranslator/route.go [hunk 24]
```diff
@@ -318,15 +307,16 @@ func (h *httpRouteConfigurationTranslator) runRoutePlugins(
 			TypedFilterConfig: typedPerFilterConfig,
 		}
 		for _, pol := range mergePolicies(pass, pols) {
-			// TODO: should we append pol.Error to errs?
-			// i.e. errs = append(errs, pol.Error)
+			// skip plugin application if we encountered any errors while constructing
+			// the policy IR.
+			if len(pol.Errors) > 0 {
+				errs = append(errs, pol.Errors...)
```
Comment (shashankram) [inline internal/kgateway/translator/irtranslator/route.go:313]: could you `logger.Error` here to indicate that policy errors were encountered and thus skipped

File: pkg/validator/validator.go (hunk 21 above)
Comment (timflannagan) [reply pkg/validator/validator.go:64]: Used primarily as a fallback mechanism for unit/envtest suites if the envoy binary doesn't exist on the filesystem. There's some multi-arch issues with the envoy-gloo binary, so we wouldn't be able to get away with instructing devs to COPY that binary from the image to run tests locally. Production environments will use the local binary.

File: hack/utils/settings/get_envoy_image.go
Comment (timflannagan) [reply hack/utils/settings/get_envoy_image.go:1]: Yep - I'll just back this change out for now and track in the parent epic.

File: internal/kgateway/translator/irtranslator/route.go (hunk 24 above)
Comment (timflannagan) [reply internal/kgateway/translator/irtranslator/route.go:313]: Seams reasonable, but I'd have to make the logging a bit more consistent. The envoyRoutes function will eventually log this aggregated/joined error message (existing code, but it also logs at debug-level which doesn't seem right), and the traffic policy plugin will additionally log this message if the Validate method fails.

File: hack/utils/settings/get_envoy_image.go
Comment (timflannagan) [reply hack/utils/settings/get_envoy_image.go:1]: Done - reverted the original commit and added a comment on the ENVOY_IMAGE Makefile variable, TODO in the validator.go to revisit this, etc in https://github.com/kgateway-dev/kgateway/pull/11553/commits/fb8ceb6f643a4e96853b082656b2a3220a994a40

File: internal/kgateway/translator/irtranslator/route.go (hunk 24 above)
Comment (shashankram) [reply internal/kgateway/translator/irtranslator/route.go:313]: As long as it is logged, I am fine with not logging it here

File: pkg/xds/bootstrap/builder.go (hunk 23 above)
Comment (timflannagan) [reply pkg/xds/bootstrap/builder.go:1]: done in https://github.com/kgateway-dev/kgateway/pull/11553/commits/7a71d41f2113089256470c49edd34787683f3f80. added a comment on the AddFilterConfig method that it assumes the staged proto.Message is valid, which I _think_ is fine since all callers would be staging via utils.MessageToAny and handle errors themselves. The alternative is we re-serialize in AddFilterConfigs and extend the signature to run an error.

File: pkg/validator/validator.go (hunk 22 above)
Comment (timflannagan) [reply pkg/validator/validator.go:1]: done in https://github.com/kgateway-dev/kgateway/pull/11553/commits/d718276cba4dc7edf776735877a1c5ce5158dc47.

**Review by shashankram — APPROVED** [review]

File: Makefile [hunk 25]
```diff
@@ -38,9 +38,7 @@ export VERSION
 
 SOURCES := $(shell find . -name "*.go" | grep -v test.go)
 
-# ATTENTION: when updating to a new major version of Envoy, check if
-# universal header validation has been enabled and if so, we expect
-# failures in `test/e2e/header_validation_test.go`.
```
Comment (lgadban) [inline Makefile:43]: is this just irrelevant at this point?

File: internal/kgateway/translator/gateway/gateway_translator_test.go [hunk 26]
```diff
@@ -861,15 +833,391 @@ var _ = DescribeTable("Basic GatewayTranslator Tests",
...
+				Expect(partiallyInvalid.Reason).To(Equal(string(gwv1.RouteReasonUnsupportedValue)))
+				Expect(partiallyInvalid.Message).To(ContainSubstring("Dropped Rule (0)"))
+				Expect(partiallyInvalid.Message).To(ContainSubstring("invalid xds configuration"))
+				Expect(partiallyInvalid.ObservedGeneration).To(Equal(int64(0)))
+			},
+		},
+		func(s *settings.Settings) {
+			s.RouteReplacementMode = settings.RouteReplacementStrict
+		}),
+	Entry("Strict Mode - Valid Structure Invalid Template (Runtime Error)",
```
Comment (lgadban) [inline internal/kgateway/translator/gateway/gateway_translator_test.go:1186]: so the strict mode even in translator tests will still call out to envoy binary for mode validate right?

**Review by lgadban — APPROVED** [review]

File: internal/kgateway/translator/gateway/gateway_translator_test.go (hunk 26 above)
Comment (timflannagan) [reply internal/kgateway/translator/gateway/gateway_translator_test.go:1186]: If the envoy binary exists on the file system, yes, otherwise falls back to the docker image for validation.

File: Makefile (hunk 25 above)
Comment (timflannagan) [reply Makefile:43]: I think so, but tagged ashish to double check whether it's still relevant.
